File: add_type.go
package main

import (
	"fmt"
	"math"
)

// Single interface combining Area and Perimeter
type ShapeOperations interface {
	Area() float64
	Perimeter() float64
}

// Existing shapes implementing ShapeOperations directly
// ---------------------------------------------------
type Circle struct{ Radius float64 }
func (c Circle) Area() float64      { return math.Pi * c.Radius * c.Radius }
func (c Circle) Perimeter() float64 { return 2 * math.Pi * c.Radius }

type Rectangle struct{ Width, Height float64 }
func (r Rectangle) Area() float64 { return r.Width * r.Height }
func (r Rectangle) Perimeter() float64 { return 2 * (r.Width + r.Height) }

// New shape added later
// ---------------------------------------------------
type Triangle struct{ Base, Height float64 }
func (t Triangle) Area() float64 { return 0.5 * t.Base * t.Height }
func (t Triangle) Perimeter() float64 {
	equalSide := math.Hypot(t.Base/2, t.Height) // For an isosceles triangle
	return t.Base + 2*equalSide
}

// Usage
// ---------------------------------------------------
func main() {
	shapes := []ShapeOperations{
		Circle{Radius: 3},                  // Original shape
		Rectangle{Width: 4, Height: 5},     // Original shape
		Triangle{Base: 6, Height: 4},       // New shape
	}

	for _, shape := range shapes {
		fmt.Printf(
			"Area: %.2f, Perimeter: %.2f\n",
			shape.Area(),
			shape.Perimeter(),
		)
	}
}

File: add_type.rs
// Define a trait with both operations
trait Shape {
    fn area(&self) -> f64;
    fn perimeter(&self) -> f64;
}

// Existing types (original code)
// --------------------------------------------
struct Circle { radius: f64 }
impl Shape for Circle {
    fn area(&self) -> f64 { std::f64::consts::PI * self.radius.powi(2) }
    fn perimeter(&self) -> f64 { 2.0 * std::f64::consts::PI * self.radius }
}

struct Rectangle { width: f64, height: f64 }
impl Shape for Rectangle {
    fn area(&self) -> f64 { self.width * self.height }
    fn perimeter(&self) -> f64 { 2.0 * (self.width + self.height) }
}

// New data type: Triangle (added later)
// --------------------------------------------
struct Triangle { base: f64, height: f64 }
impl Shape for Triangle {
    fn area(&self) -> f64 { 0.5 * self.base * self.height }
    fn perimeter(&self) -> f64 {
        let hypotenuse = (self.base / 2.0).hypot(self.height);
        self.base + 2.0 * hypotenuse
    }
}

// Usage
// --------------------------------------------
fn main() {
    let shapes: Vec<Box<dyn Shape>> = vec![
        Box::new(Circle { radius: 3.0 }),
        Box::new(Rectangle { width: 4.0, height: 5.0 }),
        Box::new(Triangle { base: 6.0, height: 4.0 }), // New type added
    ];

    for shape in shapes {
        println!(
            "Area: {:.2}, Perimeter: {:.2}",
            shape.area(),
            shape.perimeter()
        );
    }
}
